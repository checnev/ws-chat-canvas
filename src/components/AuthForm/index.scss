@use 'sass:math';
$colors: (
  rgb(255, 208, 78),
  rgb(236, 115, 128),
  rgb(119, 221, 119),
);

@function linearGradient() {
  $gradient: '';

  @for $i from 1 through length($colors) {
    @if (i == 1) {
      $gradient: nth($colors, $i) percentage($i / length($colors)) unquote(',');
    } @else if (i != length($colors)) {
      $gradient: $gradient nth($colors, $i) percentage($i / length($colors)) unquote(',');
    } @else {
      $gradient: $gradient nth($colors, $i) percentage($i / length($colors));
    }
  }

  @return $gradient;
}

.authForm {
  position: relative;

  &_fetching {
    &:before{
      content: '';
      position: absolute;
      display: block;
      width: 100%;
      height: 100%;
      z-index: 1;
      background: rgba(255, 255, 255, .6);
    }

    &:after {
      content: '';
      position: absolute;
      display: block;
      width: 60px;
      height: 60px;
      top: calc(50% - 30px);
      left: calc(50% - 30px);
      background: url('preloader.svg') center no-repeat;
      background-size: contain;
      z-index: 2;
      animation: rotate 1.5s infinite linear;
    }
  }

  &__fieldGroup {
    display: block;
    margin-bottom: 12px;
  }
}

@keyframes rotate {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
