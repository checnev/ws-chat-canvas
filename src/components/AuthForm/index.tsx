import React from 'react';
import cn from 'utils/cn';
import { Form, FormikProps } from 'formik';
import Field from 'components/common/Field';
import ColorSelector from 'components/SelectColor';
import Button from 'components/common/Button';
import './index.scss';

export interface AuthFormValues {
  name: string;
  color: string;
}

type AuthFormProps = FormikProps<AuthFormValues> & {
  isFetching: boolean;
};

class AuthForm extends React.Component<AuthFormProps> {
  render() {
    const form = cn('authForm');
    return (
      <Form className={form({ fetching: this.props.isFetching })}>
        <div className={form('fieldGroup')}>
          <Field
            name="name"
            type="text"
            label="Your Name"
            placeholder="Naruto Uzumaki"
            className={form('field')}
            onChange={this.props.handleChange}
            onBlur={this.props.handleBlur}
            value={this.props.values.name}
            invalid={Boolean(this.props.touched.name && this.props.errors.name)}
            helpMessage={this.props.errors.name}
          />
        </div>
        <div className={form('fieldGroup')}>
          <ColorSelector 
            onChange={this.props.handleChange}
            value={this.props.values.color}
            height={62.5}
          />
        </div>
        <div className={form('fieldGroup')}>
          <Button type="submit">Sign In</Button>
        </div>
      </Form>
    );
  }
}

export default AuthForm;
