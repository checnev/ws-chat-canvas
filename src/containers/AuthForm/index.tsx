import React from 'react';
import { Formik, FormikErrors, FormikProps } from 'formik';
import AuthForm, { AuthFormValues } from 'components/AuthForm';
import { UserContext } from 'context/user';
import { colors } from 'constants/colors';
import { randomNumber } from 'utils/random';

class AuthFormContainer extends React.Component {
  static contextType = UserContext;
  context!: React.ContextType<typeof UserContext>

  render() {
    return (
      <Formik
        initialValues={{
          name: '',
          color: colors[randomNumber(0, colors.length - 1)],
        }}
        onSubmit={(values: AuthFormValues) => {
          if (!this.context.isFetching) {
            const user = {
              name: values.name.trim(),
              color: values.color,
            };

            this.context.signin(user);
          }
        }}
        validate={(values: AuthFormValues) => {
          const errors: FormikErrors<AuthFormValues> = {};

          if (!values.name.trim()) {
            errors.name = 'User Name is empty!';
          }

          return errors;
        }}
      >
        {(props: FormikProps<AuthFormValues>) => <AuthForm {...props} isFetching={this.context.isFetching} />}
      </Formik>
    );
  }
}

export default AuthFormContainer;
