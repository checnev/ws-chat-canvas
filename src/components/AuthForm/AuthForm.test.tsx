import { render, unmountComponentAtNode } from 'react-dom';
import { act } from "react-dom/test-utils";
import pretty from 'pretty';
import { Formik, FormikProps } from 'formik';
import AuthForm, { AuthFormValues } from 'components/AuthForm';

let container: any = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const setUp = () => (
  <Formik
    initialValues={{
      name: 'Naruto',
      color: '#fff',
    }}
    onSubmit={(values: AuthFormValues) => {
    }}
  >
    {(props: FormikProps<AuthFormValues>) => <AuthForm {...props} />}
  </Formik>
);

it('should render AuthForm', () => {
  act(() => {
    render(setUp(), container);
  });

  expect(
    pretty(container.innerHTML)
  ).toMatchSnapshot();
});
