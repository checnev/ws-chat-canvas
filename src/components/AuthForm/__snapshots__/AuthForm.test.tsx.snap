// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AuthForm 1`] = `
"<form action=\\"#\\" class=\\"authForm\\">
  <div class=\\"authForm__fieldGroup\\"><label class=\\"input__label input__label_hasLabel\\">
      <div class=\\"input__fieldWrapper\\"><span class=\\"input__labelText\\">Your Name</span><input type=\\"text\\" class=\\"authForm__field input input_type_text\\" name=\\"name\\" placeholder=\\"Naruto Uzumaki\\" value=\\"Naruto\\"></div>
    </label></div>
  <div class=\\"authForm__fieldGroup\\">
    <fieldset class=\\"colorSelector\\"><label class=\\"colorSelector__label\\" style=\\"width: 25%;\\"><input type=\\"radio\\" class=\\"colorSelector__input\\" name=\\"color\\" value=\\"#ffd04e\\"><span class=\\"colorSelector__colorSquare\\" style=\\"background-color: rgb(255, 208, 78); height: 62.5px;\\"></span></label><label class=\\"colorSelector__label\\" style=\\"width: 25%;\\"><input type=\\"radio\\" class=\\"colorSelector__input\\" name=\\"color\\" value=\\"#ec7380\\"><span class=\\"colorSelector__colorSquare\\" style=\\"background-color: rgb(236, 115, 128); height: 62.5px;\\"></span></label><label class=\\"colorSelector__label\\" style=\\"width: 25%;\\"><input type=\\"radio\\" class=\\"colorSelector__input\\" name=\\"color\\" value=\\"#77dd77\\"><span class=\\"colorSelector__colorSquare\\" style=\\"background-color: rgb(119, 221, 119); height: 62.5px;\\"></span></label><label class=\\"colorSelector__label\\" style=\\"width: 25%;\\"><input type=\\"radio\\" class=\\"colorSelector__input\\" name=\\"color\\" value=\\"#50bade\\"><span class=\\"colorSelector__colorSquare\\" style=\\"background-color: rgb(80, 186, 222); height: 62.5px;\\"></span></label><label class=\\"colorSelector__label\\" style=\\"width: 25%;\\"><input type=\\"radio\\" class=\\"colorSelector__input\\" name=\\"color\\" value=\\"#ffe5b4\\"><span class=\\"colorSelector__colorSquare\\" style=\\"background-color: rgb(255, 229, 180); height: 62.5px;\\"></span></label><label class=\\"colorSelector__label\\" style=\\"width: 25%;\\"><input type=\\"radio\\" class=\\"colorSelector__input\\" name=\\"color\\" value=\\"#d191f9\\"><span class=\\"colorSelector__colorSquare\\" style=\\"background-color: rgb(209, 145, 249); height: 62.5px;\\"></span></label><label class=\\"colorSelector__label\\" style=\\"width: 25%;\\"><input type=\\"radio\\" class=\\"colorSelector__input\\" name=\\"color\\" value=\\"#3a9ff1\\"><span class=\\"colorSelector__colorSquare\\" style=\\"background-color: rgb(58, 159, 241); height: 62.5px;\\"></span></label><label class=\\"colorSelector__label\\" style=\\"width: 25%;\\"><input type=\\"radio\\" class=\\"colorSelector__input\\" name=\\"color\\" value=\\"#ffc2ce\\"><span class=\\"colorSelector__colorSquare\\" style=\\"background-color: rgb(255, 194, 206); height: 62.5px;\\"></span></label></fieldset>
  </div>
  <div class=\\"authForm__fieldGroup\\"><button class=\\"button button_color_primary\\" type=\\"submit\\">Sign In</button></div>
</form>"
`;
